### PR TITLE
Shavronskyy/refactor provider admin to employee

### PR DESCRIPTION
**Pull request by Shavronskyy**
#1565 
## Summary of issue
Refactor `ProviderAdmin` subrole into a completely new role - `employee`.
## Summary of change
- Replace `ProviderAdmin` with `Employee`;
- Remove IsDeputy property from `Employee` and all checks for it;
- Remove subroles;
- Remove all additional checks for subrole and helper methods for operations with subrole. Check for providerAdmin subrole replaced on check for employee role;
- Adjust [`ApplicationStatusPermissions`](https://github.com/ita-social-projects/OoS-Backend/pull/1574/files#diff-c194e7af0d5d1d6250b04f1c417155279307a416e3a87ad85a239d2c5d4b09c9) to give `Employee` all permissions of provider with `ProviderAdmin` subrole;
- Add [`EmploymentType`](https://github.com/ita-social-projects/OoS-Backend/pull/1574/files#diff-bd769bfe655cd78433c42cb8b2e0a3303347c32c66c7f99ab95dd2cd54d3faf5) enum for `Offical` model.
- Add models: 
[`Official`](https://github.com/ita-social-projects/OoS-Backend/pull/1574/files#diff-a600021825c1cfc1fafdddbf19cbe81eb3a5d3d662ee99977b285033d01d1a56) (#1532), 
[`Individual`](https://github.com/ita-social-projects/OoS-Backend/pull/1574/files#diff-4e8df3f73392e693c81a85c8c10a4075ef07ae59ec9de303c121f5bef7871520) (#1527), 
[`Position`](https://github.com/ita-social-projects/OoS-Backend/pull/1574/files#diff-65b2fcafb966ba63a0be15c14ba841b6c8fae6f33ffdb28b3c5a1ff01f52cd6c) (#1531);
- Rename at least most of ProviderAdmin variadbles to be employee...

### CODE REVIEW

File: OutOfSchool/OutOfSchool.AuthorizationServer/Services/ProfileService.cs [hunk 1]
```diff
@@ -60,7 +57,7 @@ private async Task<string> GetPermissionsForUser(User userFromLogin, string role
     {
         if (userFromLogin.Role == nameof(Role.Provider).ToLower() && userFromLogin.IsDerived)
         {
-            // ProviderAdmin set of permissions in DB excludes not allowed actions
+            // Employee set of permissions in DB excludes not allowed actions
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.AuthorizationServer/Services/ProfileService.cs]: we can remove this `if`, as `employee` is a separate role and can't be derived from provider as it was before.

File: OutOfSchool/OutOfSchool.BusinessLogic/Hubs/ChatWorkshopHub.cs [hunk 2]
```diff
@@ -252,6 +248,8 @@ private async Task<bool> UserHasRigtsForChatRoomAsync(Guid workshopId, Guid pare
 
         bool userRoleIsProvider = userRole.Equals(Role.Provider.ToString(), StringComparison.OrdinalIgnoreCase);
 
+        bool userRoleIsEmployee = userRole.Equals(Role.Employee.ToString(), StringComparison.OrdinalIgnoreCase);
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Hubs/ChatWorkshopHub.cs]: we can inject and use `ICurrentUser` interface and it's implementation to check if user is in given role.

File: OutOfSchool/OutOfSchool.BusinessLogic/Models/Changes/ProviderAdminType.cs [hunk 3]
```diff
@@ -1,11 +1 @@
...
-
-[JsonConverter(typeof(JsonStringEnumConverter))]
-public enum ProviderAdminType
-{
-    All,
-    Deputies,
-    Assistants, // workshop provider admins
-}
\ No newline at end of file
+﻿
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Models/Changes/ProviderAdminType.cs]: file is empty but not deleted?

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CurrentUserService.cs [hunk 4]
```diff
@@ -67,12 +65,14 @@ public bool HasClaim(string type, Func<string, bool>? valueComparer = null) =>
...
+        Role.Employee => IsInRole("employee"),
         _ => throw new NotImplementedException("Role not handled"),
     };
 
-    public bool IsDeputyOrProviderAdmin() =>
-        IsInRole(Role.Provider) &&
-        (IsInSubRole(Subrole.ProviderDeputy) || IsInSubRole(Subrole.ProviderAdmin));
+    public bool IsEmployeeOrProvider() => IsInRole(Role.Provider) || IsInRole(Role.Employee);
+
+    public bool isProvider() => IsInRole(Role.Provider);
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Services/CurrentUserService.cs]: the following 2 methods were not added to interface. Are they really needed then?

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs [hunk 5]
```diff
@@ -300,11 +300,11 @@ public async Task<ProviderDto> GetByUserId(string id, bool isDeputyOrAdmin = fal
 
         if (isDeputyOrAdmin)
         {
-            var providerAdmins = await providerAdminRepository.GetByFilter(p => p.UserId == id).ConfigureAwait(false);
-            var providerAdmin = providerAdmins.FirstOrDefault();
-            if (providerAdmin != null)
+            var employees = await employeeRepository.GetByFilter(p => p.UserId == id).ConfigureAwait(false);
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs:303]: Need to rename `isDeputyOrAdmin` argument so it reflects new role :)

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/ValidationService.cs [hunk 6]
```diff
@@ -42,24 +42,10 @@ public async Task<bool> UserIsWorkshopOwnerAsync(string userId, Guid workshopId,
...
-
-        if (userSubrole == Subrole.ProviderAdmin)
-        {
-            var providersAdmins = await providerAdminRepository.GetByFilter(p => p.ManagedWorkshops.Any(w => w.Id == workshopId)
-                && !p.IsDeputy
-                && p.UserId == userId).ConfigureAwait(false);
-
-            return providersAdmins.Any();
-        }
+        var providersAdmins = await employeeRepository.GetByFilter(p => p.ManagedWorkshops.Any(w => w.Id == workshopId)
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Services/ValidationService.cs]: `providersAdmins` variable naming.

File: OutOfSchool/OutOfSchool.DataAccess/Enums/Subrole.cs [hunk 7]
```diff
@@ -1,11 +1 @@
...
-
-[JsonConverter(typeof(JsonStringEnumConverter))]
-public enum Subrole
-{
-    None,
-    ProviderDeputy,
-    ProviderAdmin,
-}
\ No newline at end of file
+﻿
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.DataAccess/Enums/Subrole.cs]: looks like file was not deleted

File: OutOfSchool/OutOfSchool.DataAccess/Models/Configurations/IndividualConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,5 @@
+using OutOfSchool.Services.Models.Configurations.Base;
+
+namespace OutOfSchool.Services.Models.Configurations;
+
+internal class IndividualConfiguration : BusinessEntityConfiguration<Individual>;
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.DataAccess/Models/Configurations/IndividualConfiguration.cs]: If you are first to merge - please add unique constraint to rnokpp 😅 
```csharp
internal class IndividualConfiguration : BusinessEntityConfiguration<Individual>
{
    public override void Configure(EntityTypeBuilder<Individual> entityTypeBuilder)
    {
        entityTypeBuilder.HasIndex(i => i.Rnokpp).IsUnique();
        base.Configure(entityTypeBuilder);
    }
}
```

File: OutOfSchool/OutOfSchool.AuthorizationServer/Services/ProfileService.cs (hunk 1 above)
Comment (jimmy2jk) [reply OutOfSchool/OutOfSchool.AuthorizationServer/Services/ProfileService.cs]: done

File: OutOfSchool/OutOfSchool.BusinessLogic/Hubs/ChatWorkshopHub.cs (hunk 2 above)
Comment (jimmy2jk) [reply OutOfSchool/OutOfSchool.BusinessLogic/Hubs/ChatWorkshopHub.cs]: done

File: OutOfSchool/OutOfSchool.BusinessLogic/Models/Changes/ProviderAdminType.cs (hunk 3 above)
Comment (jimmy2jk) [reply OutOfSchool/OutOfSchool.BusinessLogic/Models/Changes/ProviderAdminType.cs]: done

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CurrentUserService.cs (hunk 4 above)
Comment (jimmy2jk) [reply OutOfSchool/OutOfSchool.BusinessLogic/Services/CurrentUserService.cs]: Removed unnecessary methods `isProvider()` and `isEmployee()`

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs (hunk 5 above)
Comment (jimmy2jk) [reply OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs:303]: Renamed to `isEmployee`

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/ValidationService.cs (hunk 6 above)
Comment (jimmy2jk) [reply OutOfSchool/OutOfSchool.BusinessLogic/Services/ValidationService.cs]: Renamed to `employees`

File: OutOfSchool/OutOfSchool.DataAccess/Enums/Subrole.cs (hunk 7 above)
Comment (jimmy2jk) [reply OutOfSchool/OutOfSchool.DataAccess/Enums/Subrole.cs]: Missed this issue while resolving conflicts. File deleted now

**Review by DmyMi — APPROVED** [review]

**Pull request merged**